// @trezor/device-authenticity/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DeviceAuthenticityConfig = exports.deviceAuthenticityConfig = exports.DeviceAuthenticityBlacklistConfig = exports.deviceAuthenticityBlacklistConfig = exports.getRandomChallenge = exports.parseCertificate = exports.parseName = exports.verifySignatureP256 = exports.verifyAuthenticityProof = void 0;
var verifyAuthenticityProof_1 = require("./verifyAuthenticityProof");
Object.defineProperty(exports, "verifyAuthenticityProof", {
  enumerable: true,
  get: function () {
    return verifyAuthenticityProof_1.verifyAuthenticityProof;
  }
});
Object.defineProperty(exports, "verifySignatureP256", {
  enumerable: true,
  get: function () {
    return verifyAuthenticityProof_1.verifySignatureP256;
  }
});
var x509certificate_1 = require("./x509certificate");
Object.defineProperty(exports, "parseName", {
  enumerable: true,
  get: function () {
    return x509certificate_1.parseName;
  }
});
Object.defineProperty(exports, "parseCertificate", {
  enumerable: true,
  get: function () {
    return x509certificate_1.parseCertificate;
  }
});
var utils_1 = require("./utils");
Object.defineProperty(exports, "getRandomChallenge", {
  enumerable: true,
  get: function () {
    return utils_1.getRandomChallenge;
  }
});
var deviceAuthenticityBlacklistConfig_1 = require("./config/deviceAuthenticityBlacklistConfig");
Object.defineProperty(exports, "deviceAuthenticityBlacklistConfig", {
  enumerable: true,
  get: function () {
    return deviceAuthenticityBlacklistConfig_1.deviceAuthenticityBlacklistConfig;
  }
});
var deviceAuthenticityBlacklistConfigTypes_1 = require("./config/deviceAuthenticityBlacklistConfigTypes");
Object.defineProperty(exports, "DeviceAuthenticityBlacklistConfig", {
  enumerable: true,
  get: function () {
    return deviceAuthenticityBlacklistConfigTypes_1.DeviceAuthenticityBlacklistConfig;
  }
});
var deviceAuthenticityConfig_1 = require("./config/deviceAuthenticityConfig");
Object.defineProperty(exports, "deviceAuthenticityConfig", {
  enumerable: true,
  get: function () {
    return deviceAuthenticityConfig_1.deviceAuthenticityConfig;
  }
});
var deviceAuthenticityConfigTypes_1 = require("./config/deviceAuthenticityConfigTypes");
Object.defineProperty(exports, "DeviceAuthenticityConfig", {
  enumerable: true,
  get: function () {
    return deviceAuthenticityConfigTypes_1.DeviceAuthenticityConfig;
  }
});
